Sources/ApolloCodegenLib/Frontend/CompilationResult+SelectionSetMerging.swift

```diff
@@ -0,0 +1,97 @@
+import Foundation
+
+protocol SelectionMergable: ScopedSelectionSetHashable {
+  var _selectionSet:CompilationResult.SelectionSet? { get }
+  func merging(_: CompilationResult.SelectionSet) -> Self
+}
+
+extension CompilationResult.SelectionSet: SelectionMergable {
+
+  var _selectionSet: CompilationResult.SelectionSet? { self }
+
+  /// Returns a `Field` with the selections of the `newSelectionSet` merged in,
+  /// removing duplicates.
+  ///
+  /// - Note: If no changes were made the same `SelectionSet` is returned.
+  func merging(
+    _ newSelectionSet: CompilationResult.SelectionSet
+  ) -> Self {
+    let selectionsToMerge = newSelectionSet.selections.filter { !selections.contains($0) }
+
+    guard !selectionsToMerge.isEmpty else { return self }
+
+    let copy = self.copy()
+    copy.selections += selectionsToMerge
+    return copy
+  }
+
+  private func copy() -> Self {
+    return Self(parentType: self.parentType,
+                selections: self.selections)
+  }
+
+}
+
+extension CompilationResult.Field: SelectionMergable {
+
+  var _selectionSet: CompilationResult.SelectionSet? { self.selectionSet }
+
+  /// Returns a `Field` with the selections of the `newSelectionSet` merged in,
+  /// removing duplicates.
+  ///
+  /// - Note: If no changes were made the same `Field` is returned.
+  func merging(_ newSelectionSet: CompilationResult.SelectionSet) -> Self {
+    guard let existingSelectionSet = selectionSet else {
+      let copy = self.copy()
+      copy.selectionSet = newSelectionSet
+      return copy
+    }
+
+    let mergedSelectionSet = existingSelectionSet.merging(newSelectionSet)
+    guard mergedSelectionSet !== existingSelectionSet else { return self }
+
+    let copy = self.copy()
+    copy.selectionSet = mergedSelectionSet
+    return copy
+  }
+
+  private func copy() -> Self {
+    return Self(
+      name: self.name,
+      alias: self.alias,
+      arguments: self.arguments,
+      type: self.type,
+      selectionSet: self.selectionSet,
+      deprecationReason: self.deprecationReason,
+      description: self.description)
+  }
+
+}
+
+extension CompilationResult.Selection: SelectionMergable {
+
+  var _selectionSet: CompilationResult.SelectionSet? {
+    switch self {
+    case let .field(selection as SelectionMergable),
+      let .inlineFragment(selection as SelectionMergable):
+      return selection._selectionSet
+
+    case let .fragmentSpread(fragment):
+      return fragment.selectionSet
+    }
+  }
+
+  func merging(_ newSelectionSet: CompilationResult.SelectionSet) -> Self {
+    switch self {
+    case let .field(field):
+      return .field(field.merging(newSelectionSet))
+
+    case let .inlineFragment(selectionSet):
+      return .inlineFragment(selectionSet.merging(newSelectionSet))
+
+    case .fragmentSpread:
+      fatalError("Selections sets should never be merged into named fragments.")
```

Comment:
This is a fatal rather than a throwing error since this basically represents an error on our part in the development of the code generator rather than anything the user can do anything about, correct?

Reply:
Correct. This would only throw if we had an invalid schema. There is a consistent pattern throughout both the codegen and the Apollo lib (in the executor and caching layer) of assuming that the schema is valid. It means we can make much cleaner code with less useless try catch block that will NEVER run.

Everything has to run through the validator in `graphql-js` before getting to this point, so we don't have to handle this case.